aim2dat/strct/ext_manipulation/add_structure.py

```diff
@@ -25,6 +26,83 @@
 cwd = os.path.dirname(__file__)
 
 
+@external_manipulation_method
+def add_structure_random(
+    structure: Structure,
+    wrap: bool = False,
+    guest_structure: Union[Structure, str] = "CH3",
+    dist_threshold: Union[float, None] = 0.8,
+    random_state: Union[float, None] = None,
+    change_label: bool = False,
+):
+    """
+    Add structure at random position and orientation.
+
+    Parameters
+    ----------
+    structure : aim2dat.strct.Structure
+        Structure to which the guest structure is added.
+    wrap : bool (optional)
+        Wrap atomic positions back into the unit cell.
+    guest_structure : str or aim2dat.strct.Structure (optional)
+        A representation of the guest structure given as a string of a functional group or molecule
+        (viable options are ``'CH3'``, ``'COOH'``, ``'H2O'``, ``'NH2'``, ``'NO2'`` or ``'OH'``), a
+        ``Structure`` object (bond direction is assumed to be the ``[-1.0, 0.0, 0.0]`` direction)
+        or the element symbol to add one single atom.
+    dist_threshold : float or None (optional)
+        Check the distances between all site pairs of the host and guest structure to ensure that
+        none of the added atoms collide.
+    random_state : float or None (optional)
+        Specify the initial random state to ensure reproducible results.
+    change_label : bool (optional)
+        Add suffix to the label of the new structure highlighting the performed manipulation.
+
+    Raises
+    ------
+    ValueError
+        Could not add guest structure, host structure seems to be too aggregated.
+    """
+    guest_strct, guest_strct_label = _check_guest_structure(guest_structure)
+
+    # In case no cell is given we would like to have the structure reasonably close:
+    if structure.cell is None:
+        threshold = 1.5 if dist_threshold is None else dist_threshold + 1.5
+        positions = np.array(structure.positions)
+        min_pos = np.amin(positions, axis=0) - threshold
+        max_pos = np.amax(positions, axis=0)
+        cell = np.zeros((3, 3))
+        for d in range(3):
+            cell[d][d] = max_pos[d] + threshold
+    else:
+        min_pos = np.zeros(3)
+        cell = np.array(structure.cell)
+
+    random.seed(a=random_state)
+    max_tries = 1000
+    for _ in range(max_tries):
+        guest_positions = np.array(guest_strct["positions"])
+        rot_v = np.array([random.random(), random.random(), random.random()])
```

Comment:
Just out of curiosity, is there a reason for not using `numpy.random.random_sample(3)`, to directly generate an array of random values (also between 0 and 1)?

Reply:
I guess, I prefer using the standard python functions when possible. In this case it is also important to have reproducible results for the tests (I guess this would also work with the numpy functions somehow...).